File: testbed_ruby/test/associations/one_to_optional_n_test.rb
require 'test/unit'
require './src-gen-umple/student_z'
require './src-gen-umple/mentor_z'
require './src-gen-umple/program_z'

module CruiseAssociations
class OneToOptionalNTest < Test::Unit::TestCase

  def test_cannotCreateNullStudent
  	assert_raise RuntimeError do #"Unable to create student due to mentor"
      StudentZ.new(99,nil)
    end
  end
  
  def test_CreateMentorWithoutStudent
    m = MentorZ.new("blah")
    assert_equal(0,m.number_of_students)
  end

  def test_CannotSetMentorNull
    m = MentorZ.new("blah")
    s = StudentZ.new(99,m)
    
    assert_equal(false,s.set_mentor(nil))
    assert_equal(m,s.get_mentor)
    assert_equal(s,m.get_student(0))
  end 
  
  def test_CreateStudentFromMentor
    m = MentorZ.new("blah")
    s = StudentZ.new(99,m)

    assert_equal(99,s.get_number)
    assert_equal(m,s.get_mentor)
    assert_equal(1,m.number_of_students)
    assert_equal(s,m.get_student(0))
  end  
  
  def test_CreateStudentFromMentorTooMany
  	assert_raise RuntimeError do #"Unable to create student due to mentor"
      m = MentorZ.new("blah")
      StudentZ.new(99,m)
      StudentZ.new(98,m)
      StudentZ.new(97,m)
      StudentZ.new(96,m)
    end
  end  
  
  def test_CannotSetMentorAlreadyAtMax
    m = MentorZ.new("blah")
    m2 = MentorZ.new("blah2")
    
    StudentZ.new(99,m)
    StudentZ.new(98,m)
    StudentZ.new(97,m)
    
    StudentZ.new(96,m2)
    StudentZ.new(95,m2)
    s = StudentZ.new(94,m2)
    assert_equal(false,s.set_mentor(m))
    
    assert_equal(3,m.number_of_students)
    assert_equal(3,m2.number_of_students)
    assert_equal(m2,s.get_mentor)
  end    

  def test_CreateStudentFromMentorJustEnough
    m = MentorZ.new("blah")
    s = StudentZ.new(99,m)
    s2 = StudentZ.new(98,m)
    s3 = StudentZ.new(97,m)
    
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
  end   
  
  def test_MaximumNumberOfStudents
    assert_equal(3,MentorZ::maximum_number_of_students)
  end
  
  def test_add_studentConstructorInformation
    m = MentorZ.new("blah")
    s = StudentZ.new(10,m)

    assert_equal(10,s.get_number)
    assert_equal(m,s.get_mentor)
    assert_equal(1,m.number_of_students)
    assert_equal(s,m.get_student(0))
  end
  
  def test_add_studentConstructorInformationTooMany
    m = MentorZ.new("blah")
    s = StudentZ.new(10,m)
    s2 = StudentZ.new(11,m)
    s3 = StudentZ.new(12,m)

    assert_equal(10,s.get_number)
    assert_equal(m,s.get_mentor)
    assert_equal(3,m.number_of_students)
    assert_equal(s,m.get_student(0))
    assert_equal(s2,m.get_student(1))
    assert_equal(s3,m.get_student(2))

    assert_raise RuntimeError do
      s4 = StudentZ.new(13,m)
    end
  end  
  
  def test_replaceMentor
    m = MentorZ.new("blah")
    m2 = MentorZ.new("blah2")

    s = StudentZ.new(123,m)
    s2 = StudentZ.new(125,m2)
    
    assert_equal(true,s.set_mentor(m2))
    
    assert_equal(m2,s.get_mentor)
    assert_equal(2,m2.number_of_students)
    assert_equal(s2,m2.get_student(0))
    assert_equal(s,m2.get_student(1))
    
    assert_equal(0,m.number_of_students)
  end 
  
  def test_addToNewMentor
    m = MentorZ.new("blah")
    m2 = MentorZ.new("blah2")

    s = StudentZ.new(123,m)
    m2.add_student(s)
    
    assert_equal(m2,s.get_mentor)
    assert_equal(s,m2.get_student(0))
    assert_equal(0,m.number_of_students)
  end  
  
  def test_removeFromExistingMentor
    m = MentorZ.new("blah")
    m2 = MentorZ.new("blah2")

    s1 = StudentZ.new(123,m)
    s2 = StudentZ.new(124,m)
    s3 = StudentZ.new(125,m)
    
    assert_equal(3,m.number_of_students)
    assert_equal(0,m2.number_of_students)
    
    m2.add_student(s1)

    assert_equal(2,m.number_of_students)
    assert_equal(s2,m.get_student(0))
    assert_equal(s3,m.get_student(1))
    
    assert_equal(1,m2.number_of_students)
    assert_equal(s1,m2.get_student(0))
    
    assert_equal(m2,s1.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
  end   

  def test_set_mentorReplacesExistingMentor
    m = MentorZ.new("blah")
    s = StudentZ.new(99,m)
    
    m2 = MentorZ.new("blah2")
    
    s.set_mentor(m2)
    assert_equal(m2,s.get_mentor)
    assert_equal(s,m2.get_student(0))
    assert_equal(0,m.number_of_students)    
  end  
  
  def test_add_studentWhenMentorHasTooMany
    m = MentorZ.new("blah")
    m2 = MentorZ.new("blah2")
    
    StudentZ.new(12,m)
    StudentZ.new(13,m)
    StudentZ.new(14,m)
    
    StudentZ.new(21,m2)
    StudentZ.new(22,m2)
    s = StudentZ.new(23,m2)
    
    assert_equal(false,m.add_student(s))

    assert_equal(3,m.number_of_students)
    assert_equal(3,m2.number_of_students)
  end
end
end

File: testbed_php/test/associations/OneToOptionalNTest.php
<?php

class OneToOptionalNTest extends UnitTestCase
{

  public function test_cannotCreateNullStudent()
  {
  	$this->expectException(new Exception("Unable to create student due to mentor"));
    new StudentZ(99,null);
  }
  

  public function test_CreateMentorWithoutStudent()
  {
    $m = new MentorZ("blah");
    $this->assertEqual(0,$m->numberOfStudents());
  }


  public function test_CannotSetMentorNull()
  {
    $m = new MentorZ("blah");
    $s = new StudentZ(99,$m);
    
    $this->assertEqual(false,$s->setMentor(null));
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($s,$m->getStudent_index(0));
  } 
  

  public function test_CreateStudentFromMentor()
  {
    $m = new MentorZ("blah");
    $s = new StudentZ(99,$m);

    $this->assertEqual(99,$s->getNumber());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual(1,$m->numberOfStudents());
    $this->assertEqual($s,$m->getStudent_index(0));
  }  
  
  public function test_CreateStudentFromMentorTooMany()
  {
  	$this->expectException(new Exception("Unable to create student due to mentor"));
    $m = new MentorZ("blah");
    new StudentZ(99,$m);
    new StudentZ(98,$m);
    new StudentZ(97,$m);
    new StudentZ(96,$m);
  }  
  

  public function test_CannotSetMentorAlreadyAtMax()
  {
    $m = new MentorZ("blah");
    $m2 = new MentorZ("blah2");
    
    new StudentZ(99,$m);
    new StudentZ(98,$m);
    new StudentZ(97,$m);
    
    new StudentZ(96,$m2);
    new StudentZ(95,$m2);
    $s = new StudentZ(94,$m2);
    $this->assertEqual(false,$s->setMentor($m));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(3,$m2->numberOfStudents());
    $this->assertEqual($m2,$s->getMentor());
  }    

  

  public function test_CreateStudentFromMentorJustEnough()
  {
    $m = new MentorZ("blah");
    $s = new StudentZ(99,$m);
    $s2 = new StudentZ(98,$m);
    $s3 = new StudentZ(97,$m);
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());
  }   
  

  public function test_MaximumNumberOfStudents()
  {
    $this->assertEqual(3,MentorZ::maximumNumberOfStudents());
  }
  

  public function test_addStudentViaConstructorInformation()
  {
    $m = new MentorZ("blah");
    $s = $m->addStudentVia(10);

    $this->assertEqual(10,$s->getNumber());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual(1,$m->numberOfStudents());
    $this->assertEqual($s,$m->getStudent_index(0));
  }
  

  public function test_addStudentViaConstructorInformationTooMany()
  {
    $m = new MentorZ("blah");
    $s = $m->addStudentVia(10);
    $s2 = $m->addStudentVia(11);
    $s3 = $m->addStudentVia(12);
    $s4 = $m->addStudentVia(13);

    $this->assertEqual(10,$s->getNumber());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($s,$m->getStudent_index(0));
    $this->assertEqual($s2,$m->getStudent_index(1));
    $this->assertEqual($s3,$m->getStudent_index(2));
    $this->assertEqual(null,$s4);
  }  
  

  public function test_replaceMentor()
  {
    $m = new MentorZ("blah");
    $m2 = new MentorZ("blah2");

    $s = $m->addStudentVia(123);
    $s2 = $m2->addStudentVia(125);
    
    $this->assertEqual(true,$s->setMentor($m2));
    
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual(2,$m2->numberOfStudents());
    $this->assertEqual($s2,$m2->getStudent_index(0));
    $this->assertEqual($s,$m2->getStudent_index(1));
    
    $this->assertEqual(0,$m->numberOfStudents());
    
  } 
  

  public function test_addToNewMentor()
  {
    $m = new MentorZ("blah");
    $m2 = new MentorZ("blah2");

    $s = $m->addStudentVia(123);
    $m2->addStudent($s);
    
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual($s,$m2->getStudent_index(0));
    $this->assertEqual(0,$m->numberOfStudents());
  }  
  

  public function test_removeFromExistingMentor()
  {
    $m = new MentorZ("blah");
    $m2 = new MentorZ("blah2");

    $s1 = $m->addStudentVia(123);
    $s2 = $m->addStudentVia(124);
    $s3 = $m->addStudentVia(125);
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(0,$m2->numberOfStudents());
    
    $m2->addStudent($s1);

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($s2,$m->getStudent_index(0));
    $this->assertEqual($s3,$m->getStudent_index(1));
    
    $this->assertEqual(1,$m2->numberOfStudents());
    $this->assertEqual($s1,$m2->getStudent_index(0));
    
    $this->assertEqual($m2,$s1->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());
  }   

  public function test_setMentorReplacesExistingMentor()
  {
    $m = new MentorZ("blah");
    $s = new StudentZ(99,$m);
    
    $m2 = new MentorZ("blah2");
    
    $s->setMentor($m2);
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual($s,$m2->getStudent_index(0));
    $this->assertEqual(0,$m->numberOfStudents());    
  }  
  

  public function test_addStudentWhenMentorHasTooMany()
  {
    $m = new MentorZ("blah");
    $m2 = new MentorZ("blah2");
    
    $m->addStudentVia(12);
    $m->addStudentVia(13);
    $m->addStudentVia(14);
    
    $m2->addStudentVia(21);
    $m2->addStudentVia(22);
    $s = $m2->addStudentVia(23);
    
    $this->assertEqual(false,$m->addStudent($s));

    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(3,$m2->numberOfStudents());
  }
}
?>
